hal/targets/hal/TARGET_STM/TARGET_STM32F4/i2c_api.c

```diff
@@ -523,7 +736,495 @@ int i2c_slave_write(i2c_t *obj, const char *data, int length)
+    if(g_transmitIT[1]) {
+        if(handle->XferCount > 0) {
+            NVIC_SetVector(I2cEventIRQs[1], (uint32_t)i2c2_irq_handler);
+        } else {
+            NVIC_SetVector(I2cEventIRQs[1], g_handler[1]); 
+        }
+    } else {
+        if(handle->XferCount > 0) {
+            NVIC_SetVector(I2cEventIRQs[1], (uint32_t)i2c2_irq_handler);
+        } else {
+            NVIC_SetVector(I2cEventIRQs[1], g_handler[1]); 
+        }
+    }
+}
+#endif
+#if defined(I2C3_BASE)
+void i2c3_irq_handler(void) {
+    I2C_HandleTypeDef *handle = &t_I2cHandle[2];
+    HAL_I2C_EV_IRQHandler(handle, g_stop[2]);
+    if(g_transmitIT[2]) {
+        if(handle->XferCount > 1) {
+            NVIC_SetVector(I2cEventIRQs[2], (uint32_t)i2c3_irq_handler);
+        } else {
+            NVIC_SetVector(I2cEventIRQs[2], g_handler[2]); 
+        }
+    } else {
+        if(handle->XferCount > 2) {
+            NVIC_SetVector(I2cEventIRQs[2], (uint32_t)i2c3_irq_handler);
+        } else {
+            NVIC_SetVector(I2cEventIRQs[2], g_handler[2]); 
+        }
+    }
+}
+#endif
+
+#if defined(I2C1_BASE) && !defined(I2C2_BASE) && !defined(I2C3_BASE)
+const uint32_t t_i2c_irq_handler[]={(uint32_t)i2c1_irq_handler};
+#elif defined(I2C1_BASE) && defined(I2C2_BASE) && !defined(I2C3_BASE)
+const uint32_t t_i2c_irq_handler[]={(uint32_t)i2c1_irq_handler, (uint32_t)i2c2_irq_handler};
+#elif defined(I2C1_BASE) && defined(I2C2_BASE) && defined(I2C3_BASE)
+const uint32_t t_i2c_irq_handler[]={(uint32_t)i2c1_irq_handler, (uint32_t)i2c2_irq_handler, (uint32_t)i2c3_irq_handler};
+#endif
+
+void i2c_transfer_asynch(i2c_t *obj, const void *tx, size_t tx_length, void *rx, size_t rx_length, uint32_t address, uint32_t stop, uint32_t handler, uint32_t event, DMAUsage hint) {
+    size_t i = 0;
+    // TODO: DMA usage is currently ignored by this way
+    (void) hint;  
+     
+    // check which use-case we have  
+    int use_tx = (tx != NULL && tx_length > 0);  
+    int use_rx = (rx != NULL && rx_length > 0); 
+  
+	  obj->tx_buff.buffer = (uint8_t *)malloc(8*tx_length);
+    for(i=0;i<tx_length;i++) {
+      ((uint8_t *)obj->tx_buff.buffer)[i] = ((uint8_t *)tx)[i];
+    }
+    obj->tx_buff.length = tx_length;
+
+    obj->tx_buff.pos = 0;  
+    obj->tx_buff.width = 8;  
+  
+    obj->rx_buff.buffer = rx;  
+    obj->rx_buff.length = rx_length;  
+    obj->rx_buff.pos = 0;  
+    obj->rx_buff.width = 8;  
+
+    g_handler[i2c_module_lookup(obj)] = handler;  
+    g_event[i2c_module_lookup(obj)] = event; 
+    g_stop_previous[i2c_module_lookup(obj)] = g_stop[i2c_module_lookup(obj)];
+    g_stop[i2c_module_lookup(obj)] = stop; 
+  
+    // register the same thunking handler for both event and error interrupt!  
+    IRQn_Type event_irq_n = I2cEventIRQs[i2c_module_lookup(obj)];  
+    IRQn_Type error_irq_n = I2cErrorIRQs[i2c_module_lookup(obj)]; 
+    if(((tx_length > 1)&&(rx_length == 0)) || ((rx_length > 1)&&(tx_length == 0)) || ((rx_length > 1)&&(tx_length > 1))) {
```

Comment:
What's the difference between those 2 handlers? what does the condition above checks? Might be worth having a comment?